source/extensions/common/wasm/null/null_vm.cc

```diff
@@ -17,6 +18,8 @@ namespace Common {
 namespace Wasm {
 namespace Null {
 
+ThreadSafeSingleton<VmGlobalStats> global_stats_;
```

Comment:
You should probably be using SingletonManager for this, though I haven't looked at the full review yet.

Reply:
Per @jmarantz you are trying to work around not having a global context, when in fact I think you are trying to get one? You should be using SingletonManager to create the WASM global context and then use that from wherever you need it. The SingletonManager is available in all such extension points you are trying to plug in to for this purpose.

Reply:
This is just a single 4 byte atomic number and passing a manager and using a factory seems like overkill considering that it doesn't have any dependencies.  How about  MUTABLE_CONSTRUCT_ON_FIRST_USE ?

Reply:
I think the reality though is that you will likely end up needing a global state, so why not just do it now the right way? Otherwise please add a TODO/comment on why this is being done this way.

Reply:
MUTABLE_CONSTRUCT_ON_FIRST_USE is one of the macros that does the right thing.

But I think the race you have is higher level than that, as commented below. I think to resolve the race you need  a mutex.

Reply:
Changed to Gauge. PTAL